bundles/mapping/mapmodule/service/map.layer.js

```diff
@@ -329,10 +346,222 @@ Oskari.clazz.define('Oskari.mapframework.service.MapLayerService',
+         * @param  {Boolean}         deleteLayer delete layer
+         * @param  {Boolean}         newLayer is new layer
+         */
+        updateGroupsLayers: function(layerId, newLayerConf, deleteLayer, newLayer) {
+            var me = this;
+            var groups = [];
+            var group = null;
+
+            if (me._layerGroups.length === 0) {
+                return;
+            }
+
+            // update or insert
+            if (!deleteLayer) {
+                var isLayerInGroup = function(groupLayers) {
+                    var founded = -1;
+                    for (var i = 0; i < groupLayers.length; i++) {
+                        var layer = groupLayers[i];
+                        if (layer.id === layerId) {
+                            founded = i;
+                            break;
+                        }
+                    }
+                    return founded;
+                };
+
+
+                var isGroup = function(groupId) {
+                    var isInGroup = jQuery.grep(newLayerConf.groups, function(group) {
+                        return group.id === groupId;
+                    });
+                    return isInGroup.length > 0;
+                };
+                var layerIndex = null;
+
+                for (var i = 0; i < me._layerGroups.length; i++) {
+                    group = me._layerGroups[i];
+
+                    // Check if layer is in main groups
+                    layerIndex = isLayerInGroup(group.layers);
+                    if (layerIndex >= 0) {
+                        if (deleteLayer) {
+                            group.layers.splice(layerIndex, 1);
+                        } else {
+                            group.layers[layerIndex] = newLayerConf;
+                        }
+                    }
+
+                    // check subgroups
+                    for (var j = 0; j < group.groups.length; j++) {
+                        var subgroup = group.groups[j];
+
+                        layerIndex = isLayerInGroup(subgroup.layers);
+                        if (layerIndex >= 0) {
+                            if (deleteLayer) {
+                                group.groups[j].layers.splice(layerIndex, 1);
+                            } else {
+                                group.groups[j].layers[layerIndex] = newLayerConf;
+                            }
+                        }
+                        // check subgroup subgroups
+                        if (subgroup.groups) {
+                            for (var k = 0; k < subgroup.groups.length; k++) {
+                                var subgroupSubgroup = subgroup.groups[k];
+                                layerIndex = isLayerInGroup(subgroupSubgroup.layers);
+                                if (layerIndex >= 0) {
+                                    if (deleteLayer) {
+                                        group.groups[j].groups[k].layers.splice(layerIndex, 1);
```

Comment:
Refactor and use recursion instead of groups.groups.groups and i j k loops. This function implementation is too long even without multiple levels of ifs and loops inside each other. Maybe break it into smaller pieces?